src/modules/gfx_autoterminal.c

```diff
@@ -0,0 +1,526 @@
+    max_index = 1;
+    shell = malloc(30 * sizeof(char));
+    strcpy(shell, "/bin/sh");
+    // read script
+    FILE *file;
+
+    file = fopen("scripts/autoterminal.sh", "r");
+    if (file) {
+        // last line endns with eof and not new line
+        int ch;
+        while ((ch = getc(file)) != EOF)
+            if ((unsigned char)ch == '\n')
+                max_index++;
+        type_buffer = malloc(max_index * sizeof(char *));
+        for (type_index = 0; type_index < max_index; type_index++) {
+            type_buffer[type_index] = malloc(max_column * 3 * sizeof(char));
+        }
+        rewind(file);
+        int first_line = 1;
+        type_index = 0;
+        while (fgets(type_buffer[type_index], max_column * 3, file) != NULL) {
+            // skip comments
+            if (type_buffer[type_index][0] == '#') {
+                if (first_line && type_buffer[0][1] == '!') {
+                    strcpy(shell, type_buffer[type_index] + 2);
+                    // each input ends with a newline
+                    int last = strlen(shell) - 1;
+                    if (shell[last] == '\n')
+                        shell[last] = '\0';
+                }
+            } else {
+                type_index++;
+                if (type_index > max_index)
+                    break;
+            }
+            first_line = 0;
+        };
+        fclose(file);
+    }
+
+    type_pos = 0;
+    max_index = type_index - 1;
+    type_index = 0;
+
+    buffer_busy = oscore_mutex_new();
+
+    buffer = malloc(max_row * max_column * sizeof(struct font_char));
+    clear_buffer();
+
+    // 2 rows
+    if (matrix_getx() < font_height * 2)
+        return 1; // not enough X to be usable
+    // 8 characters
+    if (matrix_gety() < font_width * 8)
+        return 1; // not enough Y to be usable
+    return 0;
+}
+
+void reset(int _modno) {
+    type_index = 0;
+    type_pos = 0;
+    current_row = 0;
+    current_column = 0;
+    fg = fg_default;
+    bg = bg_default;
+    clear_buffer();
+    nexttick = udate();
+}
+
+int draw(int _modno, int argc, char *argv[]) {
+    int x;
+    int y;
+    int row;
+    int column;
+    int pos = 0;
+    if (active_shell == 0) {
+        // we are through with all commands
+        if (type_index > max_index)
+            return 1;
+        // this happens right after other thread finished
+        if (type_pos == 0) { // prepare next line or exit
+            write_buffer("$ ", &current_row, &current_column);
+        }
+        if (type_pos < strlen(type_buffer[type_index])) {
+            char ch[2];
+            ch[0] = type_buffer[type_index][type_pos];
+            ch[1] = 0;
+            write_buffer(ch, &current_row, &current_column);
+            type_pos++;
+        } else {
+            current_column = 0;
+            active_shell = 1;
+            type_index++;
+            type_pos = 0;
+            oscore_task_create("shell", launch, type_buffer[type_index - 1]);
```

Comment:
You should store the thread handle and join it when it is done, otherwise you'll be leaking memory.

Reply:
`oscore_event_*` is not suitable, I don’t wanna block draw. I need some polling.
I always draw the buffer and I only want to do the autotyping if no child process is running.